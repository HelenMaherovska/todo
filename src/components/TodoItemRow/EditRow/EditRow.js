import React from "react";
import PropTypes from "prop-types";

const EditRow = props => {
  const { text, id, onToggleEdit, onEditTodo, setRef } = props;
  return (
    <tr>
      <td>
        <input
          defaultValue={text}
          type="text"
          ref={setRef}
          className="form-control todo-text"
        />
      </td>
      <td />
      <td>
        <button
          onClick={() => onEditTodo(id)}
          type="button"
          className="btn btn-success mr-2"
        >
          Save
        </button>
        <button
          onClick={onToggleEdit}
          type="button"
          className="btn btn-danger mr-2"
        >
          Cancel
        </button>
      </td>
    </tr>
  );
};

EditRow.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  onEditTodo: PropTypes.func.isRequired,
  onToggleEdit: PropTypes.func.isRequired,
  setRef: PropTypes.func.isRequired
};

export default EditRow;
