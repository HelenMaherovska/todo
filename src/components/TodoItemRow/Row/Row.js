import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

const Row = props => {
  const { id, text, completed, date, onToggleEdit, onClick, onDelete } = props;
  let todoTextClass = classNames({ "todo-text": true, completed });
  return (
    <tr>
      <td onClick={() => onClick(id)} className={todoTextClass}>
        {text}
      </td>
      <td onClick={() => onClick(id)} className={todoTextClass}>
        {date}
      </td>
      <td>
        <button
          onClick={onToggleEdit}
          type="button"
          className="btn btn-info mr-2"
        >
          Edit
        </button>
        <button
          onClick={() => onDelete(id)}
          type="button"
          className="btn btn-danger"
        >
          Delete
        </button>
      </td>
    </tr>
  );
};

Row.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onToggleEdit: PropTypes.func.isRequired
};

export default Row;
