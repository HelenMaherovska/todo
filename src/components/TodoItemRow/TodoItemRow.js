import React, { Component } from "react";
import PropTypes from "prop-types";
import Row from "components/TodoItemRow/Row/Row";
import EditRow from "components/TodoItemRow/EditRow/EditRow";
import "components/TodoItemRow/TodoItemRow.css";

class TodoItemRow extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditing: false
    };
  }
  setRef = input => {
    this.editInputRef = input;
  };

  onToggleEdit = () => {
    this.setState(prevState => ({
      isEditing: !prevState.isEditing
    }));
  };

  onEditTodo = () => {
    const { id, completed, onUpdate } = this.props;

    this.setState(prevState => ({
      isEditing: !prevState.isEditing
    }));

    const data = {
      id,
      completed,
      text: this.editInputRef.value
    };

    onUpdate(data);
  };

  render() {
    const { isEditing } = this.state;

    let todoItem = isEditing ? (
      <EditRow
        setRef={this.setRef}
        onEditTodo={this.onEditTodo}
        onToggleEdit={this.onToggleEdit}
        {...this.props}
      />
    ) : (
      <Row onToggleEdit={this.onToggleEdit} {...this.props} />
    );

    return todoItem;
  }
}

TodoItemRow.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired
};

export default TodoItemRow;
