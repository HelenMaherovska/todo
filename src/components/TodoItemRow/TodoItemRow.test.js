import React from "react";
import { shallow } from "enzyme";
import TodoItemRow from "components/TodoItemRow/TodoItemRow";
import Row from "components/TodoItemRow/Row/Row";
import EditRow from "components/TodoItemRow/EditRow/EditRow";

jest.mock("date-fns", () => () => ({ format: () => "06/18/2018 05:49:28" }));

describe("TodoItemRow", () => {
  const id = "12345";
  const text = "Learn unit testing";
  const completed = false;
  const date = "06/18/2018 05:49:28";
  const mockToggle = jest.fn();
  const mockDelete = jest.fn();
  const mockUpdate = jest.fn();

  const props = {
    id,
    text,
    completed,
    date,
    onClick: mockToggle,
    onDelete: mockDelete,
    onUpdate: mockUpdate
  };
  const todoItemRow = shallow(<TodoItemRow {...props} />);

  it("renders correctly", () => {
    expect(todoItemRow).toMatchSnapshot();
  });

  it("initilizes isEditing value in `state`", () => {
    expect(todoItemRow.state()).toEqual({ isEditing: false });
  });

  describe("default markup", () => {
    describe("isEditing value is false", () => {
      it("renders Row component", () => {
        expect(todoItemRow.find(Row)).toHaveLength(1);
      });
    });
  });
  describe("edit markup", () => {
    describe("isEditing value is true", () => {
      beforeEach(() => {
        todoItemRow.setState({ isEditing: true });
      });

      it("renders EditRow component", () => {
        expect(todoItemRow.find(EditRow)).toHaveLength(1);
      });
    });
  });
});
