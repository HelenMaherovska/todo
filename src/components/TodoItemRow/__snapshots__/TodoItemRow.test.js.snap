// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemRow renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoItemRow
    completed={false}
    date="06/18/2018 05:49:28"
    id="12345"
    onClick={[MockFunction]}
    onDelete={[MockFunction]}
    onUpdate={[MockFunction]}
    text="Learn unit testing"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "completed": false,
      "date": "06/18/2018 05:49:28",
      "id": "12345",
      "onClick": [MockFunction],
      "onDelete": [MockFunction],
      "onToggleEdit": [Function],
      "onUpdate": [MockFunction],
      "text": "Learn unit testing",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "completed": false,
        "date": "06/18/2018 05:49:28",
        "id": "12345",
        "onClick": [MockFunction],
        "onDelete": [MockFunction],
        "onToggleEdit": [Function],
        "onUpdate": [MockFunction],
        "text": "Learn unit testing",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
